### ISSUE TITLE
Support setting fields for scale domain 

### ISSUE THREAD

**Issue by kanitw** [vega/vega-lite#3818]
Consider when we want to [encode point marks with a pre-binned data](https://github.com/vega/vega-lite/issues/2912#issuecomment-392211157)

We want to encode the point position to be `bin_mid`, but set the scale domain to combine `bin_start` and `bin_end`.  Thus, I think we should extend `scale.domain` to support fields.  

Two options for the `domain` syntax:

a) `domain: {fields: ['bin_x_start', 'bin_x_end']}`
b) `domain: [{field: 'bin_x_start'}, {field: 'bin_x_end'}]

I kinda like a) better since domain as an array feels like each item are one domain value.  In this case, we want a way to combine multiple fields.

**Comment by g3o2** [comment]
Is `type` not needed in the general case?

**Comment by kanitw** [comment]
> Is type not needed in the general case?

Not sure what you mean "the general case".  But `type` in fieldDef is used to determine scale type and other properties.  Here scale type is already determined by the main field def.  So I don't think we need `type` here.

**Comment by domoritz** [comment]
If we do b, then we also need to support `domain: [{field: 'bin_x_start'}, 12]`. If we can support that easily, then we should go with b as it is more flexible.

**Comment by kanitw** [comment]
`domain: [{field: 'bin_x_start'}, 12]` is an array with mixed types (a reference to all data in a field and just one data item in a single array).   Unless you have a convincing use case, I think this adds unnecessary complexity to the system.

**Comment by domoritz** [comment]
We do support `domain: [0, 10]`, though, right? I don't think it's in any way confusing to users what's supposed to happen.

**Comment by kanitw** [comment]
`domain: [0, 10]` = `number[]` thus it isn't a mixed array.  

Currently items in the domain array are only values in the domain.  Mixing reference and specific item is confusing.  [Vega also makes their distinction clear](https://vega.github.io/vega/docs/scales/#domain).

**Comment by domoritz** [comment]
I haven't checked but doesn't Vega-Lite produce a mixed array when we merge scale domains from layers with field domains and value domains? I just feel it's an artificial restriction to the expressiveness of the language. Should we ask for a third opinion?

**Comment by kanitw** [comment]
It will make the syntax complicated and allow bad specs (e.g., `domain: [{field: 'bin_x_start'}, 12, {field: 'bin_x_end'}, 10]`. Also,  the code will be way more complicated. 

1) If you don't have a use case, I don't see why we wanna support it.    
2) If you really wanna support it, better support it in another way.

**Comment by domoritz** [comment]
is `domain: [{field: 'bin_x_start'}, 12, {field: 'bin_x_end'}, 10]` generated vega or vega-lite?

My use case is that I might want to say that the domain should be at least `-1 to 5` but if the data is outside the domain, it should be expanded. I was doing something like that in https://vega.github.io/vega/examples/pi-monte-carlo/.

```
"domain": {
        "fields": [
          [2, 4],
          {"data": "pi", "field": "value"},
          {"data": "pi_estimates", "field": "estimate"}
        ]
      },
```

**Comment by kanitw** [comment]
> is domain: [{field: 'bin_x_start'}, 12, {field: 'bin_x_end'}, 10] generated vega or vega-lite?

Vega-Lite.  (if you allow mixing types)

Using "fields" to combine them like in Vega is better (although still confusing, but at least Vega does the similar thing). The difference is that in VL there won't be `data`.

**Comment by domoritz** [comment]
`domain: {fields: ['bin_x_start', 'bin_x_end', 0, 10]}` won't work if the domain is not numeric. How about `domain: {fields: ['bin_x_start', 'bin_x_end'], values: [0, 10]}`?

**Comment by kanitw** [comment]
> domain: {fields: ['bin_x_start', 'bin_x_end'], values: [0, 10]}

This is even clearer than Vega lol.  

I think one can implement the `fields` only option first for the bin case.  The `values` is less important and can come later.

**Comment by domoritz** [comment]
The issue is that `domain: [0,10]` and `domain: {values: [0,10]}` means the same now

**Comment by kanitw** [comment]
> The issue is that domain: [0,10] and domain: {values: [0,10]} means the same now

I don't think that is an issue.  It's the same for `mark: "bar"` = `mark: {"type": "bar"}`

**Comment by domoritz** [comment]
And I'm not a fan of it ;-)

I think for mark it makes sense as it is such a common case, though.

**Comment by kanitw** [comment]
We can just support `fields` only like I suggest from the beginning too.  (That's what we should implement first anyway.)

**Comment by domoritz** [comment]
I'm okay with that but I think option b is more flexible when we think about the future and so I prefer it.

**Comment by kanitw** [comment]
Option a) is extensible too like you just proposed (in a way less problematic fashion).  

I think b) is __very bad__.

**Comment by domoritz** [comment]
But it violates the principle of not having two ways to do the same thing. Why is b very bad?

**Comment by kanitw** [comment]
And if we support `domain: [{field: 'bin_x_start'}, {field: 'bin_x_end'}, [10, 12]]`  then `domain: [[10, 12]]` = `domain: [10, 12]` -- so still two ways of doing things.

**Comment by davidwales** [comment]
Would this make it possible to set the domain based on the full data set as it exists before applying a filter transform?
I have a problem where I'm creating multiple time series charts, and I want them all to have the same maximum domain as defined by the maximum date in the original dataset.

e.g. I want to be able to do something like:

``` json
"scale": {"domainMax": {"year": 2020, "month": 9, "date": 8}}
```

but with the TimeUnit pulled directly from the original data, rather than manually specified.

**Comment by domoritz** [comment]
I think right now, you need to define a hidden mark layer to achieve this but we definitely want to support setting the scale domain based on unfiltered data as it helps with interactivity.
